File: examples/cpp/source/implicit_als/impl_als_csr_distr.cpp
/* file: impl_als_csr_distr.cpp */
/*******************************************************************************
* Copyright 2014-2016 Intel Corporation
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*     http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*******************************************************************************/

/*
!  Content:
!    C++ example of the implicit alternating least squares (ALS) algorithm in
!    the distributed processing mode.
!
!    The program trains the implicit ALS model on a training data set.
!******************************************************************************/

/**
 * <a name="DAAL-EXAMPLE-CPP-IMPLICIT_ALS_CSR_DISTRIBUTED"></a>
 * \example impl_als_csr_distr.cpp
 */

#include "daal.h"
#include "service.h"

using namespace std;
using namespace daal;
using namespace daal::algorithms::implicit_als;

/* Input data set parameters */
const size_t nBlocks = 4;

/* Number of observations in training data set blocks */
const string trainDatasetFileNames[nBlocks] =
{
    "../data/distributed/implicit_als_csr_1.csv",
    "../data/distributed/implicit_als_csr_2.csv",
    "../data/distributed/implicit_als_csr_3.csv",
    "../data/distributed/implicit_als_csr_4.csv"
};

/* Number of observations in transposed training data set blocks */
const string transposedTrainDatasetFileNames[nBlocks] =
{
    "../data/distributed/implicit_als_trans_csr_1.csv",
    "../data/distributed/implicit_als_trans_csr_2.csv",
    "../data/distributed/implicit_als_trans_csr_3.csv",
    "../data/distributed/implicit_als_trans_csr_4.csv"
};

static size_t usersPartition[nBlocks + 1];
static size_t itemsPartition[nBlocks + 1];

typedef double  algorithmFPType;    /* Algorithm floating-point type */
typedef double  dataFPType;         /* Input data floating-point type */

/* Algorithm parameters */
const size_t nUsers = 46;           /* Full number of users */
const size_t nFactors = 2;          /* Number of factors */
const size_t maxIterations = 5;     /* Number of iterations in the implicit ALS training algorithm */

services::SharedPtr<CSRNumericTable> dataTable[nBlocks];
services::SharedPtr<CSRNumericTable> transposedDataTable[nBlocks];

NumericTablePtr predictedRatings[nBlocks][nBlocks];

KeyValueDataCollectionPtr userFactorsToNodes[nBlocks];
KeyValueDataCollectionPtr itemFactorsToNodes[nBlocks];

services::SharedPtr<training::DistributedPartialResultStep4> itemsPartialResultLocal[nBlocks];
services::SharedPtr<training::DistributedPartialResultStep4> usersPartialResultLocal[nBlocks];

services::SharedPtr<training::DistributedPartialResultStep1> step1LocalResult[nBlocks];
NumericTablePtr step2MasterResult;
KeyValueDataCollectionPtr step3LocalResult[nBlocks];
KeyValueDataCollectionPtr step4LocalInput[nBlocks];

void initializeModel(size_t block);
void readData(size_t block);
void trainModel();
void testModel();
void printResults();

void computeStep1Local(size_t block, services::SharedPtr<training::DistributedPartialResultStep4> *partialResultLocal);
void computeStep2Master();
void computeStep3Local(size_t block, size_t *offsets,
                       services::SharedPtr<training::DistributedPartialResultStep4> *partialResultLocal,
                       KeyValueDataCollectionPtr *factorsToNodes);
void computeStep4Local(size_t block, services::SharedPtr<CSRNumericTable> *dataTable,
                       services::SharedPtr<training::DistributedPartialResultStep4> *partialResultLocal);

int main(int argc, char *argv[])
{
    for (size_t i = 0; i < nBlocks; i++)
    {
        readData(i);
        initializeModel(i);

        step4LocalInput[i] = KeyValueDataCollectionPtr(new KeyValueDataCollection());
    }

    computePartialModelBlocksToNode(nBlocks, dataTable, transposedDataTable, usersPartition, itemsPartition,
                userFactorsToNodes, itemFactorsToNodes);

    trainModel();

    testModel();

    printResults();

    return 0;
}

void readData(size_t block)
{
    /* Read trainDatasetFileName from a file and create a numeric table to store the input data */
    dataTable[block] = services::SharedPtr<CSRNumericTable>(
        createSparseTable<dataFPType>(trainDatasetFileNames[block]));
    /* Read trainDatasetFileName from a file and create a numeric table to store the input data */
    transposedDataTable[block] = services::SharedPtr<CSRNumericTable>(
        createSparseTable<dataFPType>(transposedTrainDatasetFileNames[block]));
}

void initializeModel(size_t block)
{
    /* Create an algorithm object to initialize the implicit ALS model with the fastCSR method */
    training::init::Distributed<step1Local, algorithmFPType, training::init::fastCSR> initAlgorithm;
    initAlgorithm.parameter.fullNUsers = nUsers;
    initAlgorithm.parameter.nFactors = nFactors;
    initAlgorithm.parameter.seed += block;
    /* Pass a training data set to the algorithm */
    initAlgorithm.input.set(training::init::data, transposedDataTable[block]);

    /* Initialize the implicit ALS model */
    initAlgorithm.compute();

    services::SharedPtr<PartialModel> partialModelLocal = initAlgorithm.getPartialResult()->get(training::init::partialModel);

    itemsPartialResultLocal[block] = services::SharedPtr<training::DistributedPartialResultStep4>(new training::DistributedPartialResultStep4());
    itemsPartialResultLocal[block]->set(training::outputOfStep4ForStep1, partialModelLocal);
}

void computeStep1Local(size_t block, services::SharedPtr<training::DistributedPartialResultStep4> *partialResultLocal)
{
    /* Create an algorithm object to perform first step of the implicit ALS training algorithm on local-node data */
    training::Distributed<step1Local> algorithm;
    algorithm.parameter.nFactors = nFactors;

    /* Set input objects for the algorithm */
    algorithm.input.set(training::partialModel, partialResultLocal[block]->get(training::outputOfStep4ForStep1));

    /* Compute partial results of the first step on local nodes */
    algorithm.compute();

    /* Get the computed partial results */
    step1LocalResult[block] = algorithm.getPartialResult();
}

void computeStep2Master()
{
    /* Create an algorithm object to perform second step of the implicit ALS training algorithm */
    training::Distributed<step2Master> algorithm;
    algorithm.parameter.nFactors = nFactors;

    /* Set the partial results of the first local step of distributed computations
       as input for the master-node algorithm */
    for (size_t i = 0; i < nBlocks; i++)
    {
        algorithm.input.add(training::inputOfStep2FromStep1, step1LocalResult[i]);
    }

    /* Compute a partial result on the master node from the partial results on local nodes */
    algorithm.compute();

    /* Get the computed partial results */
    step2MasterResult = algorithm.getPartialResult()->get(training::outputOfStep2ForStep4);
}

void computeStep3Local(size_t block, size_t *offsets, services::SharedPtr<training::DistributedPartialResultStep4> *partialResultLocal,
                       KeyValueDataCollectionPtr *factorsToNodes)
{
    /* Create an algorithm object to perform third step of the implicit ALS training algorithm on local-node data */
    training::Distributed<step3Local> algorithm;
    algorithm.parameter.nFactors = nFactors;

    services::SharedPtr<HomogenNumericTable<size_t> > offsetTable(
            new HomogenNumericTable<size_t>(&offsets[block], 1, 1));

    /* Set input objects for the algorithm */
    algorithm.input.set(training::partialModel,             partialResultLocal[block]->get(training::outputOfStep4ForStep3));
    algorithm.input.set(training::partialModelBlocksToNode, factorsToNodes[block]);
    algorithm.input.set(training::offset,                   offsetTable);

    /* Compute partial results of the third step on local nodes */
    algorithm.compute();

    /* Get the computed partial results */
    step3LocalResult[block] = algorithm.getPartialResult()->get(training::outputOfStep3ForStep4);

    /* Prepare input objects for the fourth step of the distributed algorithm */
    for (size_t i = 0; i < nBlocks; i++)
    {
        (*step4LocalInput[i])[block] = (*step3LocalResult[block])[i];
    }
}

void computeStep4Local(size_t block, services::SharedPtr<CSRNumericTable> *dataTable,
                       services::SharedPtr<training::DistributedPartialResultStep4> *partialResultLocal)
{
    /* Create an algorithm object to perform fourth step of the implicit ALS training algorithm on local-node data */
    training::Distributed<step4Local> algorithm;
    algorithm.parameter.nFactors = nFactors;

    /* Set input objects for the algorithm */
    algorithm.input.set(training::partialModels,         step4LocalInput[block]);
    algorithm.input.set(training::partialData,           dataTable[block]);
    algorithm.input.set(training::inputOfStep4FromStep2, step2MasterResult);

    /* Build the implicit ALS partial model on the local node */
    algorithm.compute();

    /* Get the local implicit ALS partial models */
    partialResultLocal[block] = algorithm.getPartialResult();
}

void trainModel()
{
    for (size_t iteration = 0; iteration < maxIterations; iteration++)
    {
        /* Update partial users factors */
        for (size_t i = 0; i < nBlocks; i++)
        {
            computeStep1Local(i, itemsPartialResultLocal);
        }
        computeStep2Master();

        for (size_t i = 0; i < nBlocks; i++)
        {
            computeStep3Local(i, itemsPartition, itemsPartialResultLocal, itemFactorsToNodes);
        }

        for (size_t i = 0; i < nBlocks; i++)
        {
            computeStep4Local(i, dataTable, usersPartialResultLocal);
        }

        /* Update partial items factors */
        for (size_t i = 0; i < nBlocks; i++)
        {
            computeStep1Local(i, usersPartialResultLocal);
        }
        computeStep2Master();

        for (size_t i = 0; i < nBlocks; i++)
        {
            computeStep3Local(i, usersPartition, usersPartialResultLocal, userFactorsToNodes);
        }

        for (size_t i = 0; i < nBlocks; i++)
        {
            computeStep4Local(i, transposedDataTable, itemsPartialResultLocal);
        }
    }
}

void testModel()
{
    for (size_t i = 0; i < nBlocks; i++)
    {
        for (size_t j = 0; j < nBlocks; j++)
        {
            /* Create an algorithm object to predict ratings based in the implicit ALS partial models */
            prediction::ratings::Distributed<step1Local> algorithm;
            algorithm.parameter.nFactors = nFactors;

            /* Set input objects for the algorithm */
            algorithm.input.set(prediction::ratings::usersPartialModel, usersPartialResultLocal[i]->get(training::outputOfStep4));
            algorithm.input.set(prediction::ratings::itemsPartialModel, itemsPartialResultLocal[j]->get(training::outputOfStep4));

            /* Predict ratings */
            algorithm.compute();

            /* Retrieve the algorithm results */
            predictedRatings[i][j] = algorithm.getResult()->get(prediction::ratings::prediction);
        }
    }
}

void printResults()
{
    for (size_t i = 0; i < nBlocks; i++)
    {
        for (size_t j = 0; j < nBlocks; j++)
        {
            cout << "Ratings for users block " << i << ", items block " << j << " :" << endl;
            printALSRatings(usersPartition[i], itemsPartition[j], predictedRatings[i][j]);
        }
    }
}

File: examples/python/source/implicit_als/impl_als_csr_distr.py
# file: impl_als_csr_distr.py
#===============================================================================
# Copyright 2014-2019 Intel Corporation.
#
# This software and the related documents are Intel copyrighted  materials,  and
# your use of  them is  governed by the  express license  under which  they were
# provided to you (License).  Unless the License provides otherwise, you may not
# use, modify, copy, publish, distribute,  disclose or transmit this software or
# the related documents without Intel's prior written permission.
#
# This software and the related documents  are provided as  is,  with no express
# or implied  warranties,  other  than those  that are  expressly stated  in the
# License.
#===============================================================================

## <a name="DAAL-EXAMPLE-PY-IMPLICIT_ALS_CSR_DISTRIBUTED"></a>
## \example impl_als_csr_distr.py

import os
import sys

import numpy as np

from daal import step1Local, step2Local, step2Master, step3Local, step4Local
import daal.algorithms.implicit_als.prediction.ratings as ratings
import daal.algorithms.implicit_als.training as training
import daal.algorithms.implicit_als.training.init as init
from daal.data_management import KeyValueDataCollection, HomogenNumericTable

utils_folder = os.path.realpath(os.path.abspath(os.path.dirname(os.path.dirname(__file__))))
if utils_folder not in sys.path:
    sys.path.insert(0, utils_folder)
from utils import createSparseTable, printALSRatings

DAAL_PREFIX = os.path.join('..', 'data')

# Input data set parameters
nBlocks = 4

# Number of observations in transposed training data set blocks
trainDatasetFileNames = [
    os.path.join(DAAL_PREFIX, 'distributed', 'implicit_als_trans_csr_1.csv'),
    os.path.join(DAAL_PREFIX, 'distributed', 'implicit_als_trans_csr_2.csv'),
    os.path.join(DAAL_PREFIX, 'distributed', 'implicit_als_trans_csr_3.csv'),
    os.path.join(DAAL_PREFIX, 'distributed', 'implicit_als_trans_csr_4.csv')
]

usersPartition = [0] * 1
usersPartition[0] = nBlocks

userOffsets = [0] * nBlocks
itemOffsets = [0] * nBlocks

# Algorithm parameters
nUsers = 46        # Full number of users
nFactors = 2       # Number of factors
maxIterations = 5  # Number of iterations in the implicit ALS training algorithm

dataTable = [0] * nBlocks
transposedDataTable = [0] * nBlocks

predictedRatings = [[0] * nBlocks for x in range(nBlocks)]

userStep3LocalInput = [0] * nBlocks
itemStep3LocalInput = [0] * nBlocks

itemsPartialResultLocal = [0] * nBlocks
usersPartialResultLocal = [0] * nBlocks

def readData(block):
    global dataTable

    # Read trainDatasetFileName from a file and create a numeric table to store the input data
    dataTable[block] = createSparseTable(trainDatasetFileNames[block])


def initializeStep1Local(block):
    global itemsPartialResultLocal
    global itemStep3LocalInput
    global userOffsets

    # Create an algorithm object to initialize the implicit ALS model with the fastCSR method
    initAlgorithm = init.Distributed(step=step1Local, method=init.fastCSR)
    initAlgorithm.parameter.fullNUsers = nUsers
    initAlgorithm.parameter.nFactors = nFactors
    initAlgorithm.parameter.seed += block
    usersPartitionArray = np.array(usersPartition, dtype=np.float64)
    usersPartitionArray.shape = (1, 1)

    initAlgorithm.parameter.partition = HomogenNumericTable(usersPartitionArray)

    # Pass a training data set to the algorithm
    initAlgorithm.input.set(init.data, dataTable[block])

    # Initialize the implicit ALS model
    partialResult = initAlgorithm.compute()
    itemStep3LocalInput[block] = partialResult.getCollection(init.outputOfInitForComputeStep3)
    userOffsets[block]         = partialResult.getCollection(init.offsets, block)
    partialModelLocal          = partialResult.getPartialModel(init.partialModel)

    itemsPartialResultLocal[block] = training.DistributedPartialResultStep4()
    itemsPartialResultLocal[block].set(training.outputOfStep4ForStep1, partialModelLocal)

    return partialResult.getTablesCollection(init.outputOfStep1ForStep2)

def initializeStep2Local(block, initStep2LocalInput):
    global transposedDataTable
    global userStep3LocalInput
    global itemOffsets
    # Create an algorithm object to initialize the implicit ALS model with the fastCSR method
    initAlgorithm = init.Distributed(step=step2Local, method=init.fastCSR)

    initAlgorithm.input.set(init.inputOfStep2FromStep1, initStep2LocalInput)

    # Initialize the implicit ALS model
    partialResult = initAlgorithm.compute()

    transposedDataTable[block] = partialResult.getTable(init.transposedData)
    userStep3LocalInput[block] = partialResult.getCollection(init.outputOfInitForComputeStep3)
    itemOffsets[block]         = partialResult.getCollection(init.offsets, block)

def initializeModel():
    initStep1LocalResult = [0] * nBlocks

    for i in range(nBlocks):
        initStep1LocalResult[i] = initializeStep1Local(i)

    initStep2LocalInput = [0] * nBlocks

    for i in range(nBlocks):
        initStep2LocalInput[i] = KeyValueDataCollection()
        for j in range(nBlocks):
            initStep2LocalInput[i][j] = initStep1LocalResult[j][i]

    for i in range(nBlocks):
        initializeStep2Local(i, initStep2LocalInput[i])


def computeStep1Local(partialResultLocal):

    # Create an algorithm object to perform first step of the implicit ALS training algorithm on local-node data
    algorithm = training.Distributed(step=step1Local)
    algorithm.parameter.nFactors = nFactors

    # Set input objects for the algorithm
    algorithm.input.set(training.partialModel, partialResultLocal.get(training.outputOfStep4ForStep1))

    # Compute partial results of the first step on local nodes
    # DistributedPartialResultStep1 class from training
    return algorithm.compute()


def computeStep2Master(step1LocalResult):

    # Create an algorithm object to perform second step of the implicit ALS training algorithm
    algorithm = training.Distributed(step=step2Master)
    algorithm.parameter.nFactors = nFactors

    # Set the partial results of the first local step of distributed computations
    # as input for the master-node algorithm
    for i in range(nBlocks):
        algorithm.input.add(training.inputOfStep2FromStep1, step1LocalResult[i])

    # Compute a partial result on the master node from the partial results on local nodes
    # DistributedPartialResultStep2 class from training
    res = algorithm.compute()
    return res.get(training.outputOfStep2ForStep4)


def computeStep3Local(offsets, partialResultLocal, step3LocalInput):

    # Create an algorithm object to perform third step of the implicit ALS training algorithm on local-node data
    algorithm = training.Distributed(step=step3Local)
    algorithm.parameter.nFactors = nFactors

    # Set input objects for the algorithm
    algorithm.input.setModel(training.partialModel, partialResultLocal.get(training.outputOfStep4ForStep3))
    algorithm.input.setCollection(training.inputOfStep3FromInit, step3LocalInput)
    algorithm.input.setTable(training.offset, offsets)

    # Compute partial results of the third step on local nodes
    # DistributedPartialResultStep3 class from training
    res = algorithm.compute()
    return res.get(training.outputOfStep3ForStep4)


def computeStep4Local(dataTable, step2MasterResult, step4LocalInput):

    # Create an algorithm object to perform fourth step of the implicit ALS training algorithm on local-node data
    algorithm = training.Distributed(step=step4Local)
    algorithm.parameter.nFactors = nFactors

    # Set input objects for the algorithm
    algorithm.input.setModels(training.partialModels, step4LocalInput)
    algorithm.input.setTable(training.partialData, dataTable)
    algorithm.input.setTable(training.inputOfStep4FromStep2, step2MasterResult)

    # Build the implicit ALS partial model on the local node
    # DistributedPartialResultStep4 class from training
    return algorithm.compute()


def trainModel():

    step1LocalResult = [0] * nBlocks
    step3LocalResult = [0] * nBlocks
    step4LocalInput  = [0] * nBlocks

    for i in range(nBlocks):
        step4LocalInput[i] = KeyValueDataCollection()

    for iteration in range(maxIterations):

        # Update partial users factors
        for i in range(nBlocks):
            step1LocalResult[i] = computeStep1Local(itemsPartialResultLocal[i])

        step2MasterResult = computeStep2Master(step1LocalResult)

        for i in range(nBlocks):
            step3LocalResult[i] = computeStep3Local(itemOffsets[i], itemsPartialResultLocal[i], itemStep3LocalInput[i])

        # Prepare input objects for the fourth step of the distributed algorithm
        for i in range(nBlocks):
            for j in range(nBlocks):
                step4LocalInput[i][j] = step3LocalResult[j][i]

        for i in range(nBlocks):
            usersPartialResultLocal[i] = computeStep4Local(transposedDataTable[i], step2MasterResult, step4LocalInput[i])

        # Update partial items factors
        for i in range(nBlocks):
            step1LocalResult[i] = computeStep1Local(usersPartialResultLocal[i])

        step2MasterResult = computeStep2Master(step1LocalResult)

        for i in range(nBlocks):
            step3LocalResult[i] = computeStep3Local(userOffsets[i], usersPartialResultLocal[i], userStep3LocalInput[i])

        # Prepare input objects for the fourth step of the distributed algorithm
        for i in range(nBlocks):
            for j in range(nBlocks):
                step4LocalInput[i][j] = step3LocalResult[j][i]

        for i in range(nBlocks):
            itemsPartialResultLocal[i] = computeStep4Local(dataTable[i], step2MasterResult, step4LocalInput[i])


def testModel(i, j):
    # Create an algorithm object to predict ratings based in the implicit ALS partial models
    algorithm = ratings.Distributed(step=step1Local, method=ratings.defaultDense)
    algorithm.parameter.nFactors = nFactors

    # Set input objects for the algorithm
    algorithm.input.set(ratings.usersPartialModel, usersPartialResultLocal[i].get(training.outputOfStep4))
    algorithm.input.set(ratings.itemsPartialModel, itemsPartialResultLocal[j].get(training.outputOfStep4))

    # Predict ratings and retrieve the algorithm results
    algorithm.compute()

    # Result class from ratings
    res = algorithm.finalizeCompute()
    return res.get(ratings.prediction)


def printResults():

    for i in range(nBlocks):
        for j in range(nBlocks):
            print("Ratings for users block {}, items block {} :".format(i, j))
            printALSRatings(userOffsets[i], itemOffsets[j], predictedRatings[i][j])

if __name__ == "__main__":
    for i in range(nBlocks):
        readData(i)

    initializeModel()

    trainModel()

    for i in range(nBlocks):
        for j in range(nBlocks):
            predictedRatings[i][j] = testModel(i, j)

    printResults()
